src/UrlGenerator.php

```diff
@@ -0,0 +1,338 @@
+     */
+    private function isRelative(string $url): bool
+    {
+        return strncmp($url, '//', 2) && strpos($url, '://') === false;
+    }
+
+
+    public function getUriPrefix(): string
+    {
+        return $this->uriPrefix;
+    }
+
+    public function setUriPrefix(string $prefix): void
+    {
+        $this->uriPrefix = $prefix;
+    }
+
+    /**
+     * Checks for any missing route parameters
+     * @param array $parts
+     * @param array $substitutions
+     * @return array with minimum required parameters if any are missing or an empty array if none are missing
+     */
+    private function missingParameters(array $parts, array $substitutions): array
+    {
+        $missingParameters = [];
+
+        // Gather required parameters
+        foreach ($parts as $part) {
+            if (is_string($part)) {
+                continue;
+            }
+
+            $missingParameters[] = $part[0];
+        }
+
+        // Check if all parameters exist
+        foreach ($missingParameters as $parameter) {
+            if (!array_key_exists($parameter, $substitutions)) {
+                // Return the parameters so they can be used in an
+                // exception if needed
+                return $missingParameters;
+            }
+        }
+
+        // All required parameters are available
+        return [];
+    }
+
+    /**
+     * @param string $name
+     * @return Route
+     */
+    private function getRoute(string $name): Route
+    {
+        if (!array_key_exists($name, $this->routes)) {
+            throw new RouteNotFoundException($name);
+        }
+
+        return $this->routes[$name];
+    }
+
+    /**
+     * @param array $parameters
+     * @param array $parts
+     * @return string
+     */
+    private function generatePath(array $parameters, array $parts): string
+    {
+        $notSubstitutedParams = $parameters;
+        $path = $this->getUriPrefix();
+
+        foreach ($parts as $part) {
+            if (is_string($part)) {
+                // Append the string
+                $path .= $part;
+                continue;
+            }
+
+            // Check substitute value with regex
+            $pattern = str_replace('~', '\~', $part[1]);
+            if (preg_match('~^' . $pattern . '$~', (string)$parameters[$part[0]]) === 0) {
+                throw new \RuntimeException(
+                    sprintf(
+                        'Parameter value for [%s] did not match the regex `%s`',
+                        $part[0],
+                        $part[1]
+                    )
+                );
+            }
+
+            // Append the substituted value
+            $path .= $parameters[$part[0]];
+            unset($notSubstitutedParams[$part[0]]);
+        }
+
+        return $path . ($notSubstitutedParams !== [] ? '?' . http_build_query($notSubstitutedParams) : '');
+    }
+
+    /**
+     * Inject queued items into the underlying router
+     */
+    private function injectItems(): void
```

Comment:
That's why I've initially implemented generator within router. If you have many URL rules, these will be injected twice for each request: once for resolving current request (UrlMatcherInterface) and again for generating URLs (UrlGeneratorInterface). Is there a way to avoid doing the same thing twice?